Index: build/dev-client.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- build/dev-client.js	(revision )
+++ build/dev-client.js	(revision )
@@ -0,0 +1,12 @@
+/* eslint-disable */
+require('eventsource-polyfill')
+var hotClient = require('webpack-hot-middleware/client?noInfo=true&reload=true')
+
+hotClient.subscribe(function (event) {
+  if (event.action === 'reload') {
+    window.location.reload()
+  }
+})
+// appID  wx5dfca9518cfe8b50
+// appsecret       4ca865d65fdb9600ae5dd0f5dabcbbca
+// gh_a8c3b581c6d1
\ No newline at end of file
Index: .idea/vcs.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- .idea/vcs.xml	(revision )
+++ .idea/vcs.xml	(revision )
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="VcsDirectoryMappings">
+    <mapping directory="$PROJECT_DIR$" vcs="Git" />
+  </component>
+</project>
\ No newline at end of file
Index: demo/node/check_sign.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- demo/node/check_sign.js	(revision )
+++ demo/node/check_sign.js	(revision )
@@ -0,0 +1,13 @@
+var sign = require('./sign.js');
+
+console.log(sign('jsapi_ticket', 'http://example.com'));
+/*
+ *something like this
+ *{
+ *  jsapi_ticket: 'jsapi_ticket',
+ *  nonceStr: '82zklqj7ycoywrk',
+ *  timestamp: '1415171822',
+ *  url: 'http://example.com',
+ *  signature: '1316ed92e0827786cfda3ae355f33760c4f70c1f'
+ *}
+ */
Index: build/webpack.config.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- build/webpack.config.js	(revision )
+++ build/webpack.config.js	(revision )
@@ -0,0 +1,43 @@
+var path = require('path')
+var webpack = require('webpack')
+var HtmlWebpackPlugin = require('html-webpack-plugin')
+module.exports = {
+    entry:{
+        app:[path.resolve(__dirname,'../src/main.js')],
+    },
+    output:{
+        path:path.resolve(__dirname,'../dist'),
+        filename:'[name].js'
+    },
+    module:{
+        rules:[
+            {
+                test: /\.js$/,
+                loader:'babel-loader',
+                include:[path.join(__dirname, '..', 'src')]
+            }
+        ]
+    },
+    devtool: '#cheap-module-eval-source-map',
+    plugins:[
+
+        new webpack.HotModuleReplacementPlugin(),
+        // new webpack.NoEmitOnErrorsPlugin(),
+        // new HtmlWebpackPlugin({
+        //     filename: path.resolve(__dirname,'../dist/index.html'),
+        //     template: 'index.html',
+        //     inject: true
+        // }),
+        new HtmlWebpackPlugin({
+            filename: 'index.html',
+            template: 'index.html',
+            inject: true
+        }),
+    ],
+    // devServer:{
+    //     contentBase: path.resolve(__dirname,'../dist/'),
+    //     historyApiFallback:true,
+    //     inline:true,
+    //     hot:true
+    // }
+}
\ No newline at end of file
Index: .idea/jsLibraryMappings.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- .idea/jsLibraryMappings.xml	(revision )
+++ .idea/jsLibraryMappings.xml	(revision )
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="JavaScriptLibraryMappings">
+    <includedPredefinedLibrary name="ECMAScript 6" />
+  </component>
+</project>
\ No newline at end of file
Index: .idea/misc.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- .idea/misc.xml	(revision )
+++ .idea/misc.xml	(revision )
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="JavaScriptSettings">
+    <option name="languageLevel" value="ES6" />
+  </component>
+</project>
\ No newline at end of file
Index: build/dev-server.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- build/dev-server.js	(revision )
+++ build/dev-server.js	(revision )
@@ -0,0 +1,178 @@
+var webpack = require('webpack')
+var path = require('path')
+var webpackConfig = require('./webpack.config')
+var express = require('express')
+var request = require('request')
+Object.keys(webpackConfig.entry).forEach(function (name) {
+    webpackConfig.entry[name] = [path.resolve(__dirname,'./dev-client.js')].concat(webpackConfig.entry[name])
+})
+var app = express()
+var compiler = webpack(webpackConfig)
+
+var devMiddleware = require('webpack-dev-middleware')(compiler, {
+    publicPath: webpackConfig.output.publicPath,
+    quiet: true
+})
+var hotMiddleware = require('webpack-hot-middleware')(compiler,{
+        log:()=>{}
+})
+compiler.plugin('compilation', function (compilation) {
+    compilation.plugin('html-webpack-plugin-after-emit', function (data, cb) {
+        console.log('------------')
+        hotMiddleware.publish({ action: 'reload' })
+        cb()
+    })
+})
+app.use(devMiddleware)
+app.use(hotMiddleware)
+
+// app.use(express.static(path.resolve(__dirname,'../')));
+app.listen('80')
+var router = express.Router();
+var AppID = 'wx5dfca9518cfe8b50';
+var AppSecret = '4ca865d65fdb9600ae5dd0f5dabcbbca';
+app.all('*', function(req, res, next) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Content-Type,Content-Length, Authorization, Accept,X-Requested-With");
+    res.header("Access-Control-Allow-Methods","PUT,POST,GET,DELETE,OPTIONS");
+    res.header("X-Powered-By",' 3.2.1')
+    if(req.method=="OPTIONS") res.send(200);/*让options请求快速返回*/
+    else  next();
+});
+// router.get('/wx_login', function(req,res, next){
+//     console.log("oauth - login")
+//
+//     // 第一步：用户同意授权，获取code
+//     var router = 'getWX';
+//     // 这是编码后的地址
+//     var return_uri = 'http%3A%2F%2F127.0.0.1%2F'+router;
+//     var scope = 'snsapi_userinfo';
+//     // request.get({
+//     //     url:'https://open.weixin.qq.com/connect/oauth2/authorize?appid='+AppID+'&redirect_uri='+return_uri+'&response_type=code&scope='+scope+'&state=STATE#wechat_redirect',
+//     //     headers:{'Access-Control-Allow-Origin':'*'}
+//     // },function(error, response, body){
+//     //     res.send(body);
+//     //     console.log(body)
+//     // })
+//     // res.header({'Access-Control-Allow-Origin':'*',
+//     //     'Origin':'https://127.0.0.1'});
+//     res.redirect('https://open.weixin.qq.com/connect/oauth2/authorize?appid='+AppID+'&redirect_uri='+return_uri+'&response_type=code&scope='+scope+'&state=STATE#wechat_redirect');
+//
+// });
+app.get('/wx_login', function(req,res, next){
+    console.log("oauth - login")
+
+    // 第一步：用户同意授权，获取code
+    var router = 'getWX';
+    // 这是编码后的地址
+    var return_uri = 'http%3A%2F%2F127.0.0.1%2F'+router;
+    var scope = 'snsapi_userinfo';
+    // request.get({
+    //     url:'https://open.weixin.qq.com/connect/oauth2/authorize?appid='+AppID+'&redirect_uri='+return_uri+'&response_type=code&scope='+scope+'&state=STATE#wechat_redirect',
+    //     headers:{'Access-Control-Allow-Origin':'*'}
+    // },function(error, response, body){
+    //     res.send(body);
+    //     console.log(body)
+    // })
+    // res.header({'Access-Control-Allow-Origin':'*',
+    //     'Origin':'https://127.0.0.1'});
+    res.redirect('https://open.weixin.qq.com/connect/oauth2/authorize?appid='+AppID+'&redirect_uri='+return_uri+'&response_type=code&scope='+scope+'&state=STATE#wechat_redirect');
+
+});
+router.get('/getWX', function(req,res, next){
+    console.log("code_return: "+req.query.code)
+    var code = req.query.code;
+    request.get(
+        {
+            url:'https://api.weixin.qq.com/sns/oauth2/access_token?appid='+AppID+'&secret='+AppSecret+'&code='+code+'&grant_type=authorization_code',
+        },
+        function(error, response, body){
+            if(response.statusCode == 200){
+
+                // 第三步：拉取用户信息(需scope为 snsapi_userinfo)
+                console.log(JSON.parse(body));
+                var data = JSON.parse(body);
+                var access_token = data.access_token;
+                var openid = data.openid;
+
+                request.get(
+                    {
+                        url:'https://api.weixin.qq.com/sns/userinfo?access_token='+access_token+'&openid='+openid+'&lang=zh_CN',
+                    },
+                    function(error, response, body){
+                        if(response.statusCode == 200){
+
+                            // 第四步：根据获取的用户信息进行对应操作
+                            var userinfo = JSON.parse(body);
+                            //console.log(JSON.parse(body));
+                            console.log('获取微信信息成功！');
+
+                            // 小测试，实际应用中，可以由此创建一个帐户
+                            res.send("\
+                                <h1>"+userinfo.nickname+" 的个人信息</h1>\
+                                <p><img src='"+userinfo.headimgurl+"' /></p>\
+                                <p>"+userinfo.city+"，"+userinfo.province+"，"+userinfo.country+"</p>\
+                            ");
+
+                        }else{
+                            console.log(response.statusCode);
+                        }
+                    }
+                );
+            }else{
+                console.log(response.statusCode);
+            }
+        }
+    );
+})
+router.get('/get_wx_access_token', function(req,res, next){
+    console.log("get_wx_access_token")
+    console.log("code_return: "+req.query.code)
+
+    return false
+    // 第二步：通过code换取网页授权access_token
+    var code = req.query.code;
+    request.get(
+        {
+            url:'https://api.weixin.qq.com/sns/oauth2/access_token?appid='+AppID+'&secret='+AppSecret+'&code='+code+'&grant_type=authorization_code',
+        },
+        function(error, response, body){
+            if(response.statusCode == 200){
+
+                // 第三步：拉取用户信息(需scope为 snsapi_userinfo)
+                console.log(JSON.parse(body));
+                var data = JSON.parse(body);
+                var access_token = data.access_token;
+                var openid = data.openid;
+
+                request.get(
+                    {
+                        url:'https://api.weixin.qq.com/sns/userinfo?access_token='+access_token+'&openid='+openid+'&lang=zh_CN',
+                    },
+                    function(error, response, body){
+                        if(response.statusCode == 200){
+
+                            // 第四步：根据获取的用户信息进行对应操作
+                            var userinfo = JSON.parse(body);
+                            //console.log(JSON.parse(body));
+                            console.log('获取微信信息成功！');
+
+                            // 小测试，实际应用中，可以由此创建一个帐户
+                            res.send("\
+                                <h1>"+userinfo.nickname+" 的个人信息</h1>\
+                                <p><img src='"+userinfo.headimgurl+"' /></p>\
+                                <p>"+userinfo.city+"，"+userinfo.province+"，"+userinfo.country+"</p>\
+                            ");
+
+                        }else{
+                            console.log(response.statusCode);
+                        }
+                    }
+                );
+            }else{
+                console.log(response.statusCode);
+            }
+        }
+    );
+});
+app.use(router)
Index: .gitignore
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- .gitignore	(revision )
+++ .gitignore	(revision )
@@ -0,0 +1,4 @@
+node_modules/
+*/node_modules/
+.idea/
+.git/
Index: demo/node/sign.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- demo/node/sign.js	(revision )
+++ demo/node/sign.js	(revision )
@@ -0,0 +1,48 @@
+var createNonceStr = function () {
+  return Math.random().toString(36).substr(2, 15);
+};
+
+var createTimestamp = function () {
+  return parseInt(new Date().getTime() / 1000) + '';
+};
+
+var raw = function (args) {
+  var keys = Object.keys(args);
+  keys = keys.sort()
+  var newArgs = {};
+  keys.forEach(function (key) {
+    newArgs[key.toLowerCase()] = args[key];
+  });
+
+  var string = '';
+  for (var k in newArgs) {
+    string += '&' + k + '=' + newArgs[k];
+  }
+  string = string.substr(1);
+  return string;
+};
+
+/**
+* @synopsis 签名算法 
+*
+* @param jsapi_ticket 用于签名的 jsapi_ticket
+* @param url 用于签名的 url ，注意必须动态获取，不能 hardcode
+*
+* @returns
+*/
+var sign = function (jsapi_ticket, url) {
+  var ret = {
+    jsapi_ticket: jsapi_ticket,
+    nonceStr: createNonceStr(),
+    timestamp: createTimestamp(),
+    url: url
+  };
+  var string = raw(ret);
+      jsSHA = require('jssha');
+      shaObj = new jsSHA(string, 'TEXT');
+  ret.signature = shaObj.getHash('SHA-1', 'HEX');
+
+  return ret;
+};
+
+module.exports = sign;
Index: demo/node/package.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- demo/node/package.json	(revision )
+++ demo/node/package.json	(revision )
@@ -0,0 +1,15 @@
+{
+  "name": "wechat_jsapi_sign",
+  "version": "1.0.0",
+  "description": "",
+  "main": "sign.js",
+  "dependencies": {
+    "jssha": "^1.5.0"
+  },
+  "devDependencies": {},
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC"
+}
Index: package.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- package.json	(revision )
+++ package.json	(revision )
@@ -0,0 +1,42 @@
+{
+  "name": "wx-node",
+  "version": "1.0.0",
+  "description": "weixin,node",
+  "main": "index.js",
+  "scripts": {
+    "dev": "node build/dev-server.js"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/luyi10year/wx-node.git"
+  },
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/luyi10year/wx-node/issues"
+  },
+  "homepage": "https://github.com/luyi10year/wx-node#readme",
+  "devDependencies": {
+    "babel-core": "^6.22.1",
+    "babel-eslint": "^7.1.1",
+    "babel-loader": "^6.2.10",
+    "babel-plugin-transform-flow-strip-types": "^6.22.0",
+    "babel-plugin-transform-runtime": "^6.22.0",
+    "babel-preset-env": "^1.3.2",
+    "babel-preset-stage-2": "^6.22.0",
+    "babel-register": "^6.22.0",
+    "eventsource-polyfill": "^0.9.6",
+    "express": "^4.16.2",
+    "flow-bin": "^0.57.3",
+    "html-webpack-plugin": "^2.30.1",
+    "opn": "^5.1.0",
+    "request": "^2.83.0",
+    "tapable": "^0.2.8",
+    "webpack": "^3.8.1",
+    "webpack-dev-middleware": "^1.12.0",
+    "webpack-dev-server": "^2.9.3",
+    "webpack-hot-middleware": "^2.20.0",
+    "webpack-merge": "^4.1.0",
+    "wechat-oauth": "^1.2.1"
+  }
+}
Index: src/main.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/main.js	(revision )
+++ src/main.js	(revision )
@@ -0,0 +1,0 @@
Index: index.html
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- index.html	(revision )
+++ index.html	(revision )
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Title</title>
+</head>
+<body>
+<button id="click">授权</button>
+
+<script src="https://res.wx.qq.com/connect/zh_CN/htmledition/js/wxLogin.js"></script>
+<script>
+ var xhr = new XMLHttpRequest();
+ document.querySelector('#click').addEventListener('click',function(){
+     xhr.open('GET', '/wx_login', true);
+//     xhr.setRequestHeader("Access-Control-Allow-Origin", "*");
+//     xhr.setRequestHeader("Access-Control-Allow-Headers", "Content-Type,Content-Length, Authorization, Accept,X-Requested-With");
+//     xhr.setRequestHeader("Access-Control-Allow-Methods","PUT,POST,GET,DELETE,OPTIONS");
+//     xhr.setRequestHeader("X-Powered-By",' 3.2.1')
+ xhr.send(null)
+ })
+
+
+</script>
+</body>
+</html>
\ No newline at end of file
